### PR TITLE
Add_intf_range

### PR DESCRIPTION
**Pull request by stevenlu99**
**- What I did**
Add interface range commands for
show interface status
show interface counters
config interface shutdown
config interface startup

To allow user to display, shutdown/startup/show counters of certain interfaces

**- How I did it**
Add support of interface in format of format like below
Ethernet2,Ethernet3,Ethernet6-8

**- How to verify it**
Build and load
run commands
show interfaces status Ethernet14,Ethernet18-20
show interfaces counters -i Ethernet14,Ethernet18-20
sudo config interface shutdown Ethernet14,Ethernet18-20
sudo config interface startup Ethernet14,Ethernet18-20

**- Previous command output (if the output of a command-line utility has changed)**

**- New command output (if the output of a command-line utility has changed)**
With interface options, it will display/shutdown/startup/show counters of user given interrfaces

### CODE REVIEW

**Comment by stevenlu99** [comment]
retest this please.

**Comment by prsunny** [comment]
can you please provide the actual output of the show commands in description?

**Comment by stevenlu99** [comment]
Sample outputs:

admin@sonic:~$ show interfaces status Ethernet8,Ethernet168-180
  Interface            Lanes    Speed    MTU          Alias    Vlan    Oper    Admin    Type    Asym PFC
-----------  ---------------  -------  -----  -------------  ------  ------  -------  ------  ----------
  Ethernet8      49,50,51,52     100G   9100   hundredGigE3  routed    down     down     N/A         N/A
Ethernet168       9,10,11,12     100G   9100  hundredGigE43  routed    down     down     N/A         N/A
Ethernet172      13,14,15,16     100G   9100  hundredGigE44  routed    down     down     N/A         N/A
Ethernet176  109,110,111,112     100G   9100  hundredGigE45  routed    down     down     N/A         N/A
Ethernet180  105,106,107,108     100G   9100  hundredGigE46  routed    down     down     N/A         N/A

admin@sonic:~$ show interfaces counters -i Ethernet8,Ethernet168-180
      IFACE    STATE    RX_OK    RX_BPS    RX_UTIL    RX_ERR    RX_DRP    RX_OVR    TX_OK    TX_BPS    TX_UTIL    TX_ERR    TX_DRP    TX_OVR
-----------  -------  -------  --------  ---------  --------  --------  --------  -------  --------  ---------  --------  --------  --------
  Ethernet8        X        0  0.00 B/s      0.00%         0         0         0        0  0.00 B/s      0.00%         0         0         0
Ethernet168        X        0  0.00 B/s      0.00%         0         0         0        0  0.00 B/s      0.00%         0         0         0
Ethernet172        X        0  0.00 B/s      0.00%         0         0         0        0  0.00 B/s      0.00%         0         0         0
Ethernet176        X        0  0.00 B/s      0.00%         0         0         0        0  0.00 B/s      0.00%         0         0         0
Ethernet180        X        0  0.00 B/s      0.00%         0         0         0        0  0.00 B/s      0.00%         0         0         0

**Comment by ben-gale** [comment]
> Sample outputs:
> 
> admin@sonic:~$ show interfaces status Ethernet8,Ethernet168-180
> Interface Lanes Speed MTU Alias Vlan Oper Admin Type Asym PFC
> 
> Ethernet8 49,50,51,52 100G 9100 hundredGigE3 routed down down N/A N/A
> Ethernet168 9,10,11,12 100G 9100 hundredGigE43 routed down down N/A N/A
> Ethernet172 13,14,15,16 100G 9100 hundredGigE44 routed down down N/A N/A
> Ethernet176 109,110,111,112 100G 9100 hundredGigE45 routed down down N/A N/A
> Ethernet180 105,106,107,108 100G 9100 hundredGigE46 routed down down N/A N/A
> 
> admin@sonic:~$ show interfaces counters -i Ethernet8,Ethernet168-180
> IFACE STATE RX_OK RX_BPS RX_UTIL RX_ERR RX_DRP RX_OVR TX_OK TX_BPS TX_UTIL TX_ERR TX_DRP TX_OVR
> 
> Ethernet8 X 0 0.00 B/s 0.00% 0 0 0 0 0.00 B/s 0.00% 0 0 0
> Ethernet168 X 0 0.00 B/s 0.00% 0 0 0 0 0.00 B/s 0.00% 0 0 0
> Ethernet172 X 0 0.00 B/s 0.00% 0 0 0 0 0.00 B/s 0.00% 0 0 0
> Ethernet176 X 0 0.00 B/s 0.00% 0 0 0 0 0.00 B/s 0.00% 0 0 0
> Ethernet180 X 0 0.00 B/s 0.00% 0 0 0 0 0.00 B/s 0.00% 0 0 0

@prsunny - does this answer your question? If so, then can this one merge please?

**Comment by prsunny** [comment]
@ben-gale , will review this. sample output is required as part of the description template

File: utilities_common/intf_filter.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# Interface filtering functions
+
+SONIC_PORT_NAME_PREFIX = "Ethernet"
+SONIC_PO_NAME_PREFIX = "PortChannel"
```
Comment (prsunny) [inline utilities_common/intf_filter.py:4]: Suggest to use LAG_NAME_PREFIX to be consistent with [swss-common](https://github.com/Azure/sonic-swss-common/blob/master/common/schema.h#L85)

File: config/main.py [hunk 2]
```diff
@@ -1770,21 +1771,28 @@ def startup(ctx, interface_name):
         if interface_name is None:
             ctx.fail("'interface_name' is None!")
 
-    if interface_name_is_valid(interface_name) is False:
-        ctx.fail("Interface name is invalid. Please enter a valid interface name!!")
+    intf_fs = parse_interface_in_filter(interface_name)
+    if len(intf_fs) == 1 and interface_name_is_valid(interface_name) is False:
```
Comment (prsunny) [inline config/main.py:1775]: Shouldn't we check for validity of each interface in the list?

File: config/main.py [hunk 3]
```diff
@@ -1770,21 +1771,28 @@ def startup(ctx, interface_name):
...
+            config_db.mod_entry("PORT", port_name, {"admin_status": "up"})
+
+    portchannel_list = config_db.get_table("PORTCHANNEL")
+    for po_name in portchannel_list.keys():
+        if po_name in intf_fs:
+            config_db.mod_entry("PORTCHANNEL", po_name, {"admin_status": "up"})
+
+    subport_list = config_db.get_table("VLAN_SUB_INTERFACE")
+    for sp_name in subport_list.keys():
+        if ps_name in intf_fs:
```
Comment (prsunny) [inline config/main.py:1791]: ps_name -> sp_name (Typo)

File: config/main.py [hunk 4]
```diff
@@ -1770,21 +1771,28 @@ def startup(ctx, interface_name):
...
+            config_db.mod_entry("VLAN_SUB_INTERFACE", po_name, {"admin_status": "up"})
```
Comment (prsunny) [inline config/main.py:1792]: po_name -> sp_name

File: config/main.py [hunk 5]
```diff
@@ -1801,19 +1809,25 @@ def shutdown(ctx, interface_name):
...
+            config_db.mod_entry("PORT", port_name, {"admin_status": "down"})
+
+    portchannel_list = config_db.get_table("PORTCHANNEL")
+    for po_name in portchannel_list.keys():
+        if po_name in intf_fs:
+            config_db.mod_entry("PORTCHANNEL", po_name, {"admin_status": "down"})
+
+    subport_list = config_db.get_table("VLAN_SUB_INTERFACE")
+    for sp_name in subport_list.keys():
+        if ps_name in intf_fs:
```
Comment (prsunny) [inline config/main.py:1828]: Please correct the names as commented above

**Comment by prsunny** [comment]
Please update doc/Command-Reference.md before the PR can be merged

File: config/main.py (hunk 2 above)
Comment (stevenlu99) [reply config/main.py:1775]: in case of len(intf_fs) == 1 meaning user only gives one single interface_name
If there are multiple interfaces in intf_fs, only existing interfaces will be taken later.

File: utilities_common/intf_filter.py (hunk 1 above)
Comment (stevenlu99) [reply utilities_common/intf_filter.py:4]: Could not find LAG_NAME_PREFIX.

File: config/main.py (hunk 4 above)
Comment (stevenlu99) [reply config/main.py:1792]: fixed

File: config/main.py (hunk 5 above)
Comment (stevenlu99) [reply config/main.py]: fixed

File: config/main.py (hunk 2 above)
Comment (prsunny) [reply config/main.py:1775]: Can you explain on where is it taken later?

File: utilities_common/intf_filter.py (hunk 1 above)
Comment (prsunny) [reply utilities_common/intf_filter.py:4]: SONIC_PO_NAME_PREFIX -> SONIC_LAG_NAME_PREFIX - I put the link only as a reference

File: config/main.py (hunk 4 above)
Comment (prsunny) [reply config/main.py:1792]: Still see it as `po_name`

File: config/main.py [hunk 6]
```diff
@@ -1801,19 +1809,25 @@ def shutdown(ctx, interface_name):
...
+
+    portchannel_list = config_db.get_table("PORTCHANNEL")
+    for po_name in portchannel_list.keys():
+        if po_name in intf_fs:
+            config_db.mod_entry("PORTCHANNEL", po_name, {"admin_status": "down"})
+
+    subport_list = config_db.get_table("VLAN_SUB_INTERFACE")
+    for sp_name in subport_list.keys():
+        if sp_name in intf_fs:
+            config_db.mod_entry("VLAN_SUB_INTERFACE", po_name, {"admin_status": "down"})
```
Comment (prsunny) [inline config/main.py:1829]: still `po_name`

File: config/main.py (hunk 2 above)
Comment (stevenlu99) [reply config/main.py:1775]: I mean if user give a single interface name say "Ethernet0" then it will be checked at line 1775. Otherwise, the interface names will be converted to a list intf_fs and will be checked at line 1781, 1786 etc.

File: config/main.py (hunk 4 above)
Comment (prsunny) [reply config/main.py:1792]: can you please provide an output of command with subinterface?

File: config/main.py (hunk 2 above)
Comment (prsunny) [reply config/main.py:1775]: @stevenlu99 , i dint mean the check in 1781. Was referring to `interface_name_is_valid`. Is that handled if you've multiple interfaces with comma seperated?

Comment (stevenlu99) [reply config/main.py:1775]: interface_name_is_valid will only check for single interface name. and that's why I had "len(intf_fs) == 1"

File: config/main.py (hunk 4 above)
Comment (stevenlu99) [reply config/main.py]: this interface range does not support subinterface. You can see for subinterface it will take as individual subinterface name as input parameter.

File: config/main.py (hunk 6 above)
Comment (stevenlu99) [reply config/main.py]: I saw it change. changed sp_name to subport, how about now?

**Comment by stevenlu99** [comment]
update doc to provide some sample outputs of new range commands
https://github.com/Azure/sonic-utilities/pull/935

File: config/main.py (hunk 6 above)
Comment (prsunny) [reply config/main.py]: I dint mean to change `sp_name`, in the` config_db.mod_entry,`  (line 1829), it was still using `po_name`. You can use the same name as used in `startup `to be consistent.

**Comment by prsunny** [comment]
Please have the command reference #935 changes in this PR itself.

File: config/main.py (hunk 6 above)
Comment (stevenlu99) [reply config/main.py]: Had tried change to sp_name but seems did not take effect. so changed to subport, and changed to sp_name again. hoping it works this time.

**Comment by stevenlu99** [comment]
How can I add a modified file into this PR? Is there an instruction?

**Comment by jleveque** [comment]
@stevenlu99: What do you mean, "add a modified file"? While on the same branch, you can edit any file, add the changed file(s), commit the changes and push. It will update the PR.

**Comment by stevenlu99** [comment]
Added sample outputs of range commands in Command-Reference.md

File: config/main.py [hunk 7]
```diff
@@ -1770,21 +1771,28 @@ def startup(ctx, interface_name):
...
+    portchannel_list = config_db.get_table("PORTCHANNEL")
+    for po_name in portchannel_list.keys():
+        if po_name in intf_fs:
+            config_db.mod_entry("PORTCHANNEL", po_name, {"admin_status": "up"})
+
+    subport_list = config_db.get_table("VLAN_SUB_INTERFACE")
+    for sp_name in subport_list.keys():
+        if sp_name in intf_fs:
+            config_db.mod_entry("VLAN_SUB_INTERFACE", sp_name, {"admin_status": "up"})
+
```
Comment (prsunny) [inline config/main.py:1793]: Please remove extra line

File: config/main.py [hunk 8]
```diff
@@ -1801,19 +1809,26 @@ def shutdown(ctx, interface_name):
...
+    for po_name in portchannel_list.keys():
+        if po_name in intf_fs:
+            config_db.mod_entry("PORTCHANNEL", po_name, {"admin_status": "down"})
+
+    subport_list = config_db.get_table("VLAN_SUB_INTERFACE")
+
+    for sp_name in subport_list.keys():
+        if sp_name in intf_fs:
+            config_db.mod_entry("VLAN_SUB_INTERFACE", sp_name, {"admin_status": "down"})
+
```
Comment (prsunny) [inline config/main.py:1831]: Please remove extra line

File: doc/Command-Reference.md [hunk 9]
```diff
@@ -3997,12 +4026,17 @@ This command displays a list of NTP peers known to the server as well as a summa
...
+     time correct to within 326797 ms
+     polling server every 1024 s
+
        remote           refid      st t when poll reach   delay   offset  jitter
   ==============================================================================
    23.92.29.245    .XFAC.          16 u    - 1024    0    0.000    0.000   0.000
   *204.2.134.164   46.233.231.73    2 u  916 1024  377    3.079    0.394   0.128
   ```
 
+
```
Comment (prsunny) [inline doc/Command-Reference.md:4039]: Please remove extra line

File: scripts/portstat [hunk 10]
```diff
@@ -319,6 +327,8 @@ Examples:
     uid = str(os.getuid())
     wait_time_in_seconds = args.period
     print_all = args.all
+    intf_fs = args.interface
+
```
Comment (prsunny) [inline scripts/portstat:331]: Please remove extra line

File: doc/Command-Reference.md [hunk 11]
```diff
@@ -3997,12 +4026,17 @@ This command displays a list of NTP peers known to the server as well as a summa
 - Example:
   ```
   admin@sonic:~$ show ntp
+  synchronised to NTP server (204.2.134.164) at stratum 3
```
Comment (prsunny) [inline doc/Command-Reference.md:4029]: Is this change intended to be part of this?

File: config/main.py (hunk 7 above)
Comment (stevenlu99) [reply config/main.py:1793]: Done

File: config/main.py (hunk 8 above)
Comment (stevenlu99) [reply config/main.py:1829]: Done

File: doc/Command-Reference.md (hunk 9 above)
Comment (stevenlu99) [reply doc/Command-Reference.md]: done

File: scripts/portstat (hunk 10 above)
Comment (stevenlu99) [reply scripts/portstat:328]: done

File: doc/Command-Reference.md (hunk 11 above)
Comment (stevenlu99) [reply doc/Command-Reference.md:4029]: Not sure, it must come from a git pull. Should I remove this or leave there?

File: config/main.py [hunk 12]
```diff
@@ -1801,19 +1807,25 @@ def shutdown(ctx, interface_name):
...
+        if port_name in intf_fs:
+            config_db.mod_entry("PORT", port_name, {"admin_status": "down"})
+
+    portchannel_list = config_db.get_table("PORTCHANNEL")
+    for po_name in portchannel_list.keys():
+        if po_name in intf_fs:
+            config_db.mod_entry("PORTCHANNEL", po_name, {"admin_status": "down"})
+
+    subport_list = config_db.get_table("VLAN_SUB_INTERFACE")
+
```
Comment (prsunny) [inline config/main.py:1825]: looks like this line got extra added

File: doc/Command-Reference.md [hunk 13]
```diff
@@ -2737,6 +2761,11 @@ This command is used for administratively bringing up the Physical interface or
   admin@sonic:~$ sudo config interface Ethernet63 startup
   ```
 
+  shutdown multiple interfaces
```
Comment (prsunny) [inline doc/Command-Reference.md:2764]: please change to startup

File: config/main.py (hunk 12 above)
Comment (stevenlu99) [reply config/main.py]: done

File: doc/Command-Reference.md (hunk 13 above)
Comment (stevenlu99) [reply doc/Command-Reference.md]: done

**Comment by ben-gale** [comment]
Can we merge this to 201911 also please? We need it to get more SPytests running also.

**Comment by prsunny** [comment]
This PR is a feature enhancement and not sure if it can be cherry-picked without conflicts. @lguohan , @abdosi  to provide recommendation here.

**Comment by ben-gale** [comment]
Please consider cherry-picking to 201911 - needed for SPytest coverage.

**Comment by ben-gale** [comment]
Thanks for adding the "Request for 201911 Branch" label Prince. What needs to happen to get this actioned? Thx.

**Comment by prsunny** [comment]
@abdosi to check

**Comment by abdosi** [comment]
@ben-gale this is enhancement and not sure we want to take it in 201911 now.

cc
@rlhui @lguohan

**Comment by ben-gale** [comment]
OK - so be it. We will move our focus for SPytest to the Master branch.

Thanks.

File: config/main.py [hunk 14]
```diff
@@ -1770,21 +1771,26 @@ def startup(ctx, interface_name):
         if interface_name is None:
```
Comment (judyjoseph) [inline config/main.py:1771]: if get_interface_naming_mode() == "alias":
        interface_name = interface_alias_to_name(interface_name)

@stevenlu99 If the interface naming mode is "alias" mode, this change will break the command ? as the API  interface_alias_to_name() don't handle interface range still ?
